js/analyzer.js

```diff
@@ -53,94 +89,251 @@ Analyzer.prototype.abortQueue = function(){
+    if (yst_preProcessor.__store.wordcount > 100) {
+        var keywordDensity = this.keywordDensityCheck();
+        var rating = this.keywordDensityRating(keywordDensity);
+        return {name: "keywordDensity", result: {keywordDensity: keywordDensity.toFixed(1)}, rating: rating};
+    } else {
+        return {name: "keywordDensity", result: null, rating: null};
+    }
+};
+
+/**
+ * checks and returns the keyword density
+ * @returns {number}
+ */
+Analyzer.prototype.keywordDensityCheck = function(){
+    var keywordMatches = yst_stringHelper.matchString(yst_preProcessor.__store.cleanText,[this.config.keyword]);
+    var keywordDensity = 0;
+    if ( keywordMatches !== null ) {
+        var keywordCount = keywordMatches.length;
+        keywordDensity = (keywordCount / yst_preProcessor.__store.wordcount - (keywordCount - 1 * keywordCount)) * 100;
     }
-    return result;
+    return keywordDensity;
 };
+
+/**
+ * rates the keyword density against given number
+ * @param keywordDensity
+ * @returns {number}
+ */
+Analyzer.prototype.keywordDensityRating = function(keywordDensity){
+    var rating;
+    switch (true) {
+        case (keywordDensity < 1):
+            rating =  4;
+            break;
+        case (keywordDensity > 4.5):
+            rating = -50;
+            break;
+        default:
+            rating = 9;
+            break;
+    }
+    return rating;
+};
+
 /**
  * checks if keywords appear in subheaders of stored cleanTextSomeTags text.
  * @returns resultObject
  */
-Analyzer.prototype.subheaderChecker = function() {
-
-    //regex for headers
-    var headers = yst_preProcessor._store.cleanTextSomeTags.match(/<h([1-6])(?:[^>]+)?>(.*?)<\/h\1>/g);
-    if(headers === null){
-        result = {name: 'subHeadings', result: null, rating: 7};
-    }else {
-        var foundInHeader = 0;
-        for (var i = 0; i < headers.length; i++) {
-            var formattedHeaders = this.stripKeywords(headers[i]);
-            if (formattedHeaders.match(new RegExp(this.config.keyword, 'g')) || headers[i].match(new RegExp(this.config.keyword, 'g'))) {
+Analyzer.prototype.subHeadings = function() {
+    var matches = yst_preProcessor.__store.cleanTextSomeTags.match(/<h([1-6])(?:[^>]+)?>(.*?)<\/h\1>/g);
+    foundInHeader = this.subHeadingsCheck(matches);
+    return this.subHeadingsRating(foundInHeader, matches);
+};
+/**
+ * subHeadings checker to check if keyword is present in given headings.
+ * @param matches
+ * @returns {number}
+ */
+Analyzer.prototype.subHeadingsCheck = function(matches){
+    var foundInHeader;
+    if (matches === null){
+        foundInHeader = -1;
+    }else{
+        foundInHeader = 0;
+        for (var i = 0; i < matches.length; i++) {
+            var formattedHeaders = yst_stringHelper.replaceString(matches[i], this.config.wordsToRemove);
+            if (formattedHeaders.match(new RegExp(this.config.keyword, "g")) || matches[i].match(new RegExp(this.config.keyword, "g"))) {
```

Comment:
Please assign `new RegExp(this.config.keyword, "g")` to a var and please that var outside the for-lus.